File: libraries/kafka-client-py/kafka_client_py/handler.py
import asyncio
from collections.abc import Awaitable, Callable

from ._message import ConsumedMessage
from .attempt_outcome import (
    AttemptOutcome,
    HandlerRaisedNonRetryable,
    HandlerRaisedPoison,
    HandlerRaisedRetryable,
    HandlerSucceeded,
)
from .dedupe import DedupeStore
from .dedupe_gate import DedupeGate, EventIdExtractor
from .dlq_publisher import DLQPublisher
from .errors import PoisonError
from .in_process_loop_state import InProcessLoopState
from .retry_context import RetryContext
from .retry_policy import RetryPolicy
from .retry_publisher import RetryPublisher
from .terminal_router import TerminalRouter

UserHandler = Callable[[ConsumedMessage], Awaitable[None]]


_IN_PROCESS_BACKOFF_SECONDS_PER_ATTEMPT = 0.1
_IN_PROCESS_BACKOFF_CEILING_SECONDS = 1.0


class MessageHandler:
    def __init__(
        self,
        user_handler: UserHandler,
        *,
        policy: RetryPolicy,
        retry_publisher: RetryPublisher,
        dlq_publisher: DLQPublisher,
        dedupe: DedupeStore | None = None,
        event_id: EventIdExtractor | None = None,
    ) -> None:
        self._user_handler = user_handler
        self._retry_policy = policy
        self._dedupe_gate = DedupeGate(
            dedupe_store=dedupe,
            event_id_extractor=event_id,
        )
        self._terminal_router = TerminalRouter(
            retry_policy=policy,
            retry_publisher=retry_publisher,
            dlq_publisher=dlq_publisher,
        )

    async def handle(self, message: ConsumedMessage) -> None:
        if await self._dedupe_gate.already_processed(message):
            return

        retry_context = RetryContext.from_message(message)
        last_retryable_exception = await self._run_in_process_attempts(
            message,
            retry_context,
        )

        if last_retryable_exception is None:
            return
        else:
            await self._terminal_router.route_terminal_failure(
                message,
                last_exception=last_retryable_exception,
                retry_context=retry_context,
                in_process_attempts_made=self._retry_policy.max_in_process_attempts,
            )

    async def _run_in_process_attempts(
        self,
        message: ConsumedMessage,
        retry_context: RetryContext,
    ) -> BaseException | None:
        loop_state = InProcessLoopState()

        for attempt_number in range(1, self._retry_policy.max_in_process_attempts + 1):
            attempt_outcome = await self._invoke_user_handler_once(message)
            await attempt_outcome.apply(
                loop_state=loop_state,
                terminal_router=self._terminal_router,
                message=message,
                retry_context=retry_context,
                attempt_number=attempt_number,
            )
            if loop_state.is_complete:
                return None
            if attempt_number < self._retry_policy.max_in_process_attempts:
                await self._sleep_between_in_process_attempts(attempt_number)

        return loop_state.last_retryable_exception

    async def _invoke_user_handler_once(self, message: ConsumedMessage) -> AttemptOutcome:
        try:
            await self._user_handler(message)
            return HandlerSucceeded()
        except PoisonError as poison_exception:
            return HandlerRaisedPoison(exception=poison_exception)
        except BaseException as raised_exception:
            if self._retry_policy.is_retryable(raised_exception):
                return HandlerRaisedRetryable(exception=raised_exception)
            return HandlerRaisedNonRetryable(exception=raised_exception)

    @staticmethod
    async def _sleep_between_in_process_attempts(attempt_number: int) -> None:
        await asyncio.sleep(
            min(
                _IN_PROCESS_BACKOFF_SECONDS_PER_ATTEMPT * attempt_number,
                _IN_PROCESS_BACKOFF_CEILING_SECONDS,
            )
        )

File: libraries/kafka-client-go/consumer/handler.go
package consumer

import (
	"context"
	"errors"
	"log/slog"
	"time"

	kafkaclient "github.com/power-finance/kafka-client-go"
	"github.com/power-finance/kafka-client-go/consumer/dedupe"
	"github.com/power-finance/kafka-client-go/publisher"
)

type UserHandler func(
	ctx context.Context,
	message kafkaclient.ConsumedMessage,
) error

const (
	inProcessBackoffPerAttempt = 100 * time.Millisecond
	inProcessBackoffCeiling    = 1 * time.Second
)

type MessageHandlerConfig struct {
	Policy         RetryPolicy
	RetryPublisher *publisher.RetryPublisher
	DLQPublisher   *publisher.DLQPublisher
	DedupeStore    dedupe.Store
	EventID        dedupe.EventIDExtractor
	Logger         *slog.Logger
}

type attemptOutcome int

const (
	outcomeProcessed attemptOutcome = iota
	outcomeRoutedToDLQ
	outcomeRetryableExhausted
)

type MessageHandler struct {
	userHandler    UserHandler
	retryPolicy    RetryPolicy
	dedupeGate     *dedupe.Gate
	terminalRouter *TerminalRouter
	logger         *slog.Logger
	sleep          func(ctx context.Context, duration time.Duration) error
}

func NewMessageHandler(userHandler UserHandler, config MessageHandlerConfig) *MessageHandler {
	logger := config.Logger
	if logger == nil {
		logger = slog.Default()
	}

	return &MessageHandler{
		userHandler: userHandler,
		retryPolicy: config.Policy,
		dedupeGate: dedupe.NewGate(
			config.DedupeStore,
			config.EventID,
			config.Logger,
		),
		terminalRouter: NewTerminalRouter(
			config.Policy,
			config.RetryPublisher,
			config.DLQPublisher,
			config.Logger,
		),
		logger: logger,
		sleep:  sleepUnlessCancelled,
	}
}

func (h *MessageHandler) Handle(ctx context.Context, message kafkaclient.ConsumedMessage) error {
	alreadyProcessed, dedupeErr := h.dedupeGate.AlreadyProcessed(ctx, message)
	if dedupeErr != nil {
		return dedupeErr
	} else if alreadyProcessed {
		return nil
	}

	retryContext := RetryContextFromMessage(message)
	outcome, lastRetryableError, runErr := h.runInProcessAttempts(ctx, message, retryContext)
	if runErr != nil {
		return runErr
	}

	switch outcome {
	case outcomeProcessed:
		return h.markProcessedBestEffort(ctx, message)
	case outcomeRoutedToDLQ:
		return nil
	case outcomeRetryableExhausted:
		return h.terminalRouter.RouteTerminalFailure(
			ctx,
			message,
			lastRetryableError,
			retryContext,
			h.retryPolicy.MaxInProcessAttempts,
		)
	}

	return nil
}

// markProcessedBestEffort records a handled event for dedupe; a mark failure is
// logged not returned, since the handler succeeded and redelivery would reprocess it.
func (h *MessageHandler) markProcessedBestEffort(ctx context.Context, message kafkaclient.ConsumedMessage) error {
	if markErr := h.dedupeGate.MarkProcessed(ctx, message); markErr != nil {
		h.logger.WarnContext(ctx, "kafka.dedupe.mark_failed", slog.String("error", markErr.Error()))
	}

	return nil
}

func (h *MessageHandler) runInProcessAttempts(
	ctx context.Context,
	message kafkaclient.ConsumedMessage,
	retryContext RetryContext,
) (outcome attemptOutcome, lastRetryableError error, terminalError error) {
	var lastRetryable error

	for attemptNumber := 1; attemptNumber <= h.retryPolicy.MaxInProcessAttempts; attemptNumber++ {
		handlerError := h.userHandler(ctx, message)
		if handlerError == nil {
			return outcomeProcessed, nil, nil
		} else if isShutdownInProgress(ctx, handlerError) {
			return outcomeRoutedToDLQ, nil, handlerError
		}

		switch {
		case errors.Is(handlerError, kafkaclient.ErrPoison):
			routeError := h.terminalRouter.RouteImmediateFailure(
				ctx, message, handlerError, retryContext, attemptNumber, "poison",
			)
			return outcomeRoutedToDLQ, nil, routeError

		case h.retryPolicy.IsRetryable(handlerError):
			lastRetryable = handlerError
			isLastAttempt := attemptNumber == h.retryPolicy.MaxInProcessAttempts
			if !isLastAttempt {
				backoff := inProcessBackoffForAttempt(attemptNumber)
				if sleepError := h.sleep(ctx, backoff); sleepError != nil {
					return outcomeRoutedToDLQ, nil, sleepError
				}
			}

		default:
			routeError := h.terminalRouter.RouteImmediateFailure(
				ctx, message, handlerError, retryContext, attemptNumber, "non_retryable",
			)
			return outcomeRoutedToDLQ, nil, routeError
		}
	}

	return outcomeRetryableExhausted, lastRetryable, nil
}

func isShutdownInProgress(ctx context.Context, handlerError error) bool {
	return ctx.Err() != nil || errors.Is(handlerError, context.Canceled)
}

func inProcessBackoffForAttempt(attemptNumber int) time.Duration {
	backoff := inProcessBackoffPerAttempt * time.Duration(attemptNumber)
	return min(backoff, inProcessBackoffCeiling)
}

func sleepUnlessCancelled(ctx context.Context, duration time.Duration) error {
	if duration <= 0 {
		return nil
	}

	timer := time.NewTimer(duration)
	defer timer.Stop()

	select {
	case <-ctx.Done():
		return ctx.Err()
	case <-timer.C:
		return nil
	}
}
